contracts/BidderRegistry.sol

```diff
@@ -170,7 +180,29 @@ contract BidderRegistry is IBidderRegistry, Ownable, ReentrancyGuard {
 
         providerAmount[provider] += amtMinusFee;
 
-        emit FundsRetrieved(bidder, amount);
+        // TODO(@ckartik): Ensure we throughly test this flow
+        BidPayment[commitmentDigest].state = State.Withdrawn;
+        BidPayment[commitmentDigest].bidAmt = 0;
+
+        emit FundsRetrieved(commitmentDigest, amt);
+    }
+
+    /**
+     * @dev Return funds to a bidder's prepay (only callable by the pre-confirmations contract).
+     * @dev reenterancy not necessary but still putting here for precaution
+     * @param bidID is the Bid ID that allows us to identify the bid, and prepayment
+     */
+    function returnFunds(bytes32 bidID) external nonReentrant onlyPreConfirmationEngine() {
+        BidState memory bidState = BidPayment[bidID];
+        require(bidState.state == State.PreConfirmed, "The bid was not preconfirmed");
```

Comment:
No, this is a require statement, so it enforces the invariant before allowing control to continue down the code.
In the case of the returnFunds code, we want the state to be in a PreConfirmed state to ensure this is the first time the we're attempting to retrieve the funds.

If the state was UnPreconfirmed, this would be the zero value indicating no such entry for a Bid with BidID was actually present in the mapping. if it was withdrawn, than we don't want reentrancy so we avoid in that situation as well.